Webtransport over HTTP/3

PatrikTrefil opened the pull request:
Implementation of [WebTransport over HTTP/3 (draft version 12)](https://datatracker.ietf.org/doc/html/draft-ietf-webtrans-http3-12) using `System.Net.Http` and `System.Net.Quic`.

Conceptual documentation is available in [dotnet/docs PR](https://github.com/PatrikTrefil/docs-webtransport/pull/1).

## Summary

  This PR adds client-side support for [WebTransport over HTTP/3 (draft 12)](https://datatracker.ietf.org/doc/html/draft-ietf-webtrans-http3-12) to the .NET runtime.

It introduces client-side support for WebTransport over HTTP/3 to the .NET runtime by introducing a new `System.Net.WebTransport` API and extending `System.Net.Http` with the HTTP/3 functionality
  required for WebTransport session establishment.

  ## Included in this PR

  - new client-side WebTransport API:
    - `ClientWebTransportSession`
    - `WebTransportSession`
    - `WebTransportStream`
    - supporting options, enums, exceptions, and error handling
  - HTTP/3 changes in `System.Net.Http` required for WebTransport session establishment
  - session and stream operations, including open/accept, graceful close, abort handling, and session limit updates
  - capsule-based control handling for session-level signaling
  - unit and functional tests covering session establishment, stream behavior, close semantics, configuration limits, subprotocol negotiation, and HTTP/3 Extended CONNECT integration

  ## Current scope

  This PR implements the core client-side session/stream model for WebTransport over HTTP/3.

  It does not include server-side support.

  ## Missing for full WebTransport over HTTP/3 support

  The implementation is not yet feature-complete. The main missing pieces are:

  - WebTransport datagrams are not implemented, which depends on QUIC datagram support ([issue](https://github.com/dotnet/runtime/issues/53533))
  - `RESET_STREAM_AT` support is not yet available in `System.Net.Quic`, which is required by the WebTransport over HTTP/3 spec
  - HTTP/3 keep-alive support in `System.Net.Http` is still needed for long-lived idle sessions ([issue](https://github.com/dotnet/runtime/issues/87478))
  - (some optional protocol features, such as prioritization and keying material exporters, are not implemented)

  ## Related documentation

  Conceptual documentation is being developed separately in: [PatrikTrefil/docs-webtransport#1](https://github.com/PatrikTrefil/docs-webtransport/pull/1)

CarnaViire commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http3/Frames/Http3ErrorCode.cs:
```diff
@@ -106,5 +106,18 @@ internal enum Http3ErrorCode : long
         /// </summary>
         QPackDecoderStreamError = 0x202,
+        /// <summary>
+        /// WEBTRANSPORT_BUFFERED_STREAM_REJECTED (0x3994bd84)
```
The name seems to be `WT_BUFFERED_STREAM_REJECTED` now? https://www.ietf.org/archive/id/draft-ietf-webtrans-http3-13.html#section-9.5-2.2.1

PatrikTrefil replied:
Sticking to naming from draft version 12.

CarnaViire commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http3/Frames/Http3ErrorCode.cs:
```diff
@@ -106,5 +106,18 @@ internal enum Http3ErrorCode : long
+        /// </summary>
+        WebTransportBufferedStreamRejected = 0x3994bd84,
+        /// <summary>
+        /// WEBTRANSPORT_SESSION_GONE (0x170d7b68)
```
Same (`WT_SESSION_GONE`) https://www.ietf.org/archive/id/draft-ietf-webtrans-http3-13.html#section-9.5-2.10.1

PatrikTrefil replied:
Sticking to naming from draft version 12.

CarnaViire commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http3/Frames/Http3ErrorCode.cs:
```diff
@@ -106,5 +106,18 @@ internal enum Http3ErrorCode : long
+        /// of all of the streams associated with the session with this error code.
+        /// </summary>
+        WebtransportSessionGone = 0x170d7b68,
     }
```
`WT_APPLICATION_ERROR` range technically seems to belong here too? https://www.ietf.org/archive/id/draft-ietf-webtrans-http3-13.html#section-9.5-4.2.1
We can't add them here, obviously, but I think it is worth mentioning them in a comment, for the sake of potential future additions.

PatrikTrefil replied:
I grouped the enum values into a region and added a comment. Again using naming from draft version 12.

CarnaViire commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http3/Http3SettingType.cs:
```diff
@@ -36,10 +36,18 @@ internal enum Http3SettingType : long
+        EnableConnect = 0x8,
+
+        /// <summary>
+        /// The SETTINGS_WEBTRANSPORT_MAX_SESSIONS
```
I guess all of these have a shortened name now in draft-13: this is now [`SETTINGS_WT_MAX_SESSIONS`](https://www.ietf.org/archive/id/draft-ietf-webtrans-http3-13.html#section-9.2); and there is a bunch of additional settings there, too, e.g. [`SETTINGS_WT_INITIAL_MAX_STREAMS_UNI`](https://www.ietf.org/archive/id/draft-ietf-webtrans-http3-13.html#SETTINGS_WT_INITIAL_MAX_STREAMS_UNI)

PatrikTrefil replied:
All of the settings are now defined in the enum.

CarnaViire commented on src/libraries/Common/tests/System/Net/Http/Http3SettingType.cs:
```diff
@@ -0,0 +1,59 @@
+
+namespace System.Net.Test.Common
+{
+        public enum Http3SettingType : long
```
_**TL;DR:** I wonder if this should remain as it was, as a single `SettingId` enum... But you can leave it like this for now, if you'd like, esp. given it's test-only._

BTW: https://www.rfc-editor.org/rfc/rfc9114#section-11.2.2-2

> While this registry is separate from the "HTTP/2 Settings" registry defined in [[HTTP/2](https://www.rfc-editor.org/rfc/rfc9114#RFC9113)], it is preferable that the assignments parallel each other. If an entry is present in only one registry, every effort SHOULD be made to avoid assigning the corresponding value to an unrelated operation. Expert reviewers MAY reject unrelated registrations that would conflict with the same value in the corresponding registry.

I guess this is the reason of a single `SettingId` enum for both H/2 and H/3 in the current main here (in tests).

More than that -- even though there _is_ a setting with the same semantics and value, but different _names_ for H/2 and H/3, the current implementation explicitly reuses the H/2 version of the name in H/3 connection -- I'd guess for the sake of similarity/common structure of the connection implementations:

H/2: [SETTINGS_MAX_HEADER_LIST_SIZE (0x06)](https://www.rfc-editor.org/rfc/rfc9113.html#section-6.5.2-2.12.1) (default: unlimited)

H/3: [SETTINGS_MAX_FIELD_SECTION_SIZE (0x06)](https://www.rfc-editor.org/rfc/rfc9114.html#section-7.2.4.1-2.2.1) (default: unlimited)

[Http3Connection.cs:](https://github.com/dotnet/runtime/blob/a531de630aeae9dd048a7db22721043e9b5cf2ff/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs#L38-L40)

> ```
>     // Server-advertised SETTINGS_MAX_FIELD_SECTION_SIZE
>     // https://www.rfc-editor.org/rfc/rfc9114.html#section-7.2.4.1-2.2.1
>     private uint _maxHeaderListSize = uint.MaxValue; // Defaults to infinite
> ```

P.S.: H/2 [defines](https://www.rfc-editor.org/rfc/rfc9113.html#section-6.5.2-2.12.1) the meaning of its "MAX_HEADER_LIST_SIZE"... as "the maximum _field section_ size that the sender is prepared to accept" 🤦‍♀️

CarnaViire commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http3/Http3SettingType.cs:
```diff
@@ -36,10 +36,18 @@ internal enum Http3SettingType : long
+        /// Indicates that the specified HTTP/3 endpoint is WebTransport-capable and the number of concurrent sessions it is willing to receive.
+        /// </summary>
+        /// <seealso href="https://datatracker.ietf.org/doc/html/draft-ietf-webtrans-http3-12#section-9.2-2"/>
+        WebTransportMaxSessions = 0xc671706a,
```
I know that it was in this place before, but I think this should be reordered with the H3_DATAGRAM value (`0x_00ff_d277`), which seems to be less than this one (`0x_c671_706a`)?

PatrikTrefil replied:
Fixed the ordering of all setting types.

CarnaViire commented on src/libraries/Common/src/System/Net/Http/aspnetcore/Http3/Http3SettingType.cs:
```diff
@@ -3,7 +3,7 @@
 namespace System.Net.Http
 {
-    internal enum Http3SettingType : long
+    public enum Http3SettingType : long
```
Can you pls remind me why this needed to be public?
Note that this is shared code with ASP.NET Core, so it will affect their API surface, too.

PatrikTrefil replied:
There used to be a build problem, I can't remember why. But, now it the build is fine with `internal`, so I changed it back to the original. Good catch!

CarnaViire commented on src/libraries/Common/tests/System/Net/Http/Http3LoopbackStream.cs:
```diff
@@ -57,15 +57,15 @@ public async Task SendUnidirectionalStreamTypeAsync(long streamType)
+            foreach (Http3SettingsEntry setting in settingsEntries)
             {
-                bytesWritten += EncodeHttpInteger((int)setting.SettingId, buffer.AsSpan(bytesWritten));
+                bytesWritten += EncodeHttpInteger((long)setting.SettingId, buffer.AsSpan(bytesWritten));
```
I'd guess, it should have used the extracted `VariableLengthIntegerHelper` now, and the implementation in this file should be removed? (I assume they _should_ be identical, but I didn't check -- since it's not removed yet, it isn't as easy to see within the PR)

CarnaViire commented on src/libraries/System.Net.Http/ref/System.Net.Http.cs:
```diff
@@ -272,6 +272,8 @@ public enum HttpRequestError
         InvalidResponse = 9,
         ResponseEnded = 10,
         ConfigurationLimitExceeded = 11,
+        MissingExtendedConnectManager = 12,
```
This is a very basic HTTP version-agnostic enum. It feels like it can go without the additions.... e.g. "MissingExtendedConnectManager" can go under "ExtendedConnectNotSupported", and depending on what the validation includes, "ServerSettingsValidationFailed" can go maybe under "HttpProtocolError", or also by "ExtendedConnectNotSupported"? 🤔 and the exact reason can be distinguished by the message and/or the inner exception

I'll take a look how this is used here, and I'll also check what happens in similar cases in WS and whether there are any other precedents. In general, this is on the border between HTTP protocol, and an "application protocol" = protocol built on top of HTTP, so we just need to double-check that the app and transport levels are properly distinguished, and the exceptions are surfaced from the correct level.

CarnaViire reviewed (commented):
First batch

PatrikTrefil committed:
test for WebTransportStream rejects

PatrikTrefil committed:
dispose array buffer when rejecting stream

PatrikTrefil committed:
fix implementation of receiving streams for closed sessions in MsQuicWebTransportExtendedConnectManager

PatrikTrefil committed:
add missing usings in Webtransport tests

PatrikTrefil committed:
add tracing to WebTransportStream.QuicExceptionHandler

PatrikTrefil committed:
fix Http3Connection.OpenOutboundStreamAsync

PatrikTrefil committed:
more WebTransport tests

PatrikTrefil committed:
improve code readability and tracing in MsQuicWebTransportSession

PatrikTrefil committed:
move test to more appropriate test class

PatrikTrefil committed:
remove unused using

PatrikTrefil committed:
give method better name

PatrikTrefil committed:
fix tests for closing pending streams after drain/goaway in WebTransport

before the tests were not guaranteed to have a pending stream - the stream could have been opened after the session was already closed.

PatrikTrefil committed:
docs for OpenMorePendingStreamsThanAllowed

PatrikTrefil committed:
refactor WebTransport session establishment tests and reorganize todos

PatrikTrefil committed:
improve organization of todos in WebTransport tests

PatrikTrefil committed:
remove finished todo

PatrikTrefil committed:
update HttpRequestError enum in ref

PatrikTrefil committed:
add CompleteWrites and WriteAsync method to WebTransportStream

PatrikTrefil committed:
todos for later

PatrikTrefil committed:
add missing <param /> tags to docstrings of WebTransportSession methods

PatrikTrefil committed:
change the API of session establishment

make it more similar to QuicConnection.ConnectAsync

PatrikTrefil committed:
improve existing tests and add more tests for WebTransportStream one-sided closing

PatrikTrefil committed:
fix WebTransportSessionCreationOptionsTests

PatrikTrefil committed:
add tests for WebTransportSessionCreationOptions

PatrikTrefil committed:
more documentation for WebTransportSession

PatrikTrefil committed:
add test for disposing of objects multiple times

PatrikTrefil committed:
multiple aborts of WebTransportStream test

PatrikTrefil committed:
refactor WTSession disposal and merge dispose and close methods

PatrikTrefil committed:
fix ref of System.Net.Http

PatrikTrefil committed:
add default stream error code to WebTransport

PatrikTrefil committed:
remove fixed todos

It's impossible to add a comment to the Dispose method, because the overload without parameters that the users see is defined in the parent class Stream.

PatrikTrefil committed:
fix WT stream

accidentally commited a Dispose call

PatrikTrefil committed:
add docs for exception during WTStream.Abort

PatrikTrefil committed:
fix handling of errors during WTSession.OpenOutboundStreamAsync

PatrikTrefil committed:
add test for aborting stream with invalid error code

PatrikTrefil committed:
remove shared http message invoker from ClientWebTransportSession

PatrikTrefil committed:
fix WebTransportStream test

The code had an accidental cast from long to uint.

PatrikTrefil committed:
fix Http3RequestStream handling of unsuccessful extended CONNECT requests

PatrikTrefil committed:
fix handling of nonsuccessful extended CONNECT requests in ClientWebTranpsortSession

PatrikTrefil committed:
add WebTransportSession establishment tests

PatrikTrefil committed:
refactor WebTransportLoopbackServer

PatrikTrefil committed:
better readability in WebTransport code

PatrikTrefil committed:
get rid of dependency on HttpClientHandlerTestBase in WebTransport tests

PatrikTrefil committed:
improve WebTransport tracing

PatrikTrefil committed:
improving wording of docstring

PatrikTrefil committed:
add handling of close handshake failure in MsQuicWebTransportSession

PatrikTrefil committed:
remove comments about ObjectDisposedException from WebTransportSession

DisposeAsync and CloseAsync have been merged, so now we only throw WebTransportException

PatrikTrefil committed:
assert session is fully closed after calling CloseAsync(long, string, CancellationToken) in tests

PatrikTrefil committed:
remove already fixed todos and unused using

PatrikTrefil committed:
rm todo that won't get done

decided not to bother with this task

PatrikTrefil committed:
fix bug with OperationCancelledException not being thrown in MsQuicWebTransportSession

PatrikTrefil committed:
improve implementation of WebTransportStream

- the default implementation of read/write operations are not optimal. We forward everything to the underlying streams.
- the call forwards also fix exception behavior - we were throwing NotSupportedException instead of InvalidOperationException in many cases

PatrikTrefil committed:
improve WebTransportStream API

prefer using properties over fields

PatrikTrefil committed:
split files in WebTransport

PatrikTrefil committed:
remove already fixed todo

PatrikTrefil committed:
fix formatting in capsule protocol code

PatrikTrefil committed:
move error messages to Strings.resx in WebTransport

PatrikTrefil committed:
fix disposing of test classes in WebTransport

PatrikTrefil committed:
disable auto redirects for default http client in tests

In tests we assume auto redirects are disabled. This will become important when we write tests for redirects.

PatrikTrefil committed:
Fix release of WebTransport session after requests with status code != 200

PatrikTrefil committed:
fix redirect handling in ClientWebTransportSession and improve naming and docs of exceptions

PatrikTrefil committed:
test for relative uri in WebTransportSessionCreationOptions

PatrikTrefil committed:
tests for redirects during WebTransport session establishment

PatrikTrefil committed:
more tests for WebTransportStream

PatrikTrefil committed:
fix exception handling in MsQuicWebTransportStream

PatrikTrefil committed:
fix WebTransportSessionEstablishmentTests

PatrikTrefil committed:
improve documentation of MsQuicWebTransportStream

PatrikTrefil committed:
update Strings.resx with better naming

PatrikTrefil committed:
extract values to constants in ClientWebTransportSession

PatrikTrefil committed:
fix default stream error codes in WebTransport

PatrikTrefil committed:
rm todo that won't get done

PatrikTrefil committed:
add tests for CanWrite/Read properties of WebTransportStream

PatrikTrefil committed:
code formatting and cleanup

PatrikTrefil committed:
more WebTransportStream tests and cleanup

PatrikTrefil committed:
fix disposal/cleanup of WebTransport stream/session

PatrikTrefil committed:
fix WebTransportStream tests

PatrikTrefil committed:
add more tests for WebTransportStream

PatrikTrefil committed:
fix ConcatenatedStream exception type thrown for writes

PatrikTrefil committed:
add test to WebTransportSesisonTest

PatrikTrefil committed:
remove already fixed todo

PatrikTrefil committed:
unify exception behavior in WebTransportStream

PatrikTrefil committed:
improve code readability in MsQuicWebTransportStream

PatrikTrefil committed:
add test for WebTransportSession

PatrikTrefil committed:
refactor WebTransportServerSession in tests

PatrikTrefil committed:
better variable naming in WT tests

PatrikTrefil committed:
add test for WebTransportSession

See removed todo for test description.

PatrikTrefil committed:
fix exception behavior of MsQuicWebTransportStream

PatrikTrefil committed:
add test for keep alive mechanism to WebTransport

PatrikTrefil committed:
add more comments and add proper object disposal to WebTransport tests

PatrikTrefil committed:
rm todo that won't get done

After further consideration, test with a proxy seemed unnecessary.

PatrikTrefil committed:
rm already fixed todo

PatrikTrefil committed:
add test to WebTransport

PatrikTrefil committed:
add session creation tests for WebTransport

PatrikTrefil committed:
fix operation cancellation handling in MsQuicWebTransportStream

PatrikTrefil committed:
fix bug in MsQuicWebTransportSession

We need the DisposeAsync to finish only after all of the resources are actually released, so we had to remove a performance optimization.

PatrikTrefil committed:
add missing await using

PatrikTrefil committed:
fix MsQuicWebTransportStream operation cancellation

PatrikTrefil committed:
optimize cleanup of MsQuicWebTransportSession

PatrikTrefil committed:
improve WebTransportSessionCloseTests

Assert more things to make sure the reaction is what we expect it to be.

PatrikTrefil committed:
more tests and cleanup todo in WebTransport

PatrikTrefil committed:
remove already fixed todo

PatrikTrefil committed:
use locks instead of monitor

PatrikTrefil committed:
fix race condition in WebTransportSessionCloseTests

PatrikTrefil committed:
fix race condition in WebTransportSessionCloseTests

We used GOAWAY to close a session, but that resulted in two possible QuicErrors beign thrown - StreamAborted and ConnectionAborted. This change makes it so that we always get StreamAborted.

PatrikTrefil committed:
remove typo in code

PatrikTrefil committed:
WebTransportSessionTestHelper cleanup

PatrikTrefil committed:
MsQuicWebTransportSession bytes sent tracking

PatrikTrefil committed:
refactor WebTransportStreamTests

PatrikTrefil committed:
stream count limits in WebTransport

PatrikTrefil committed:
fix CapsuleConsumer.Dispose implementation

PatrikTrefil committed:
remove already fixed todo

PatrikTrefil committed:
add parameters for HTTP version in WebTransport

PatrikTrefil committed:
remove already fixed todo

The solution described in the todo was not implemented, instead we wait until a stream becomes available to mirror the behavior of System.Net.Quic.

PatrikTrefil committed:
improve dispose implementation in WebTransportSession

Now we follow the dispose pattern guideline.

PatrikTrefil committed:
split WebTransportStreamTests into two files

PatrikTrefil committed:
remove ClientWebTransportSession.IsSupported because it does not make sense anymore

After we removed the default http message invoker, this property does not make sense anymore.

PatrikTrefil committed:
rename test method

PatrikTrefil committed:
remove todos that won't get fixed

PatrikTrefil committed:
small improvements to WebTransport test code

PatrikTrefil committed:
better documentation for WebTransportStream

PatrikTrefil committed:
improve comments in Http3ErrorCode

PatrikTrefil committed:
revert change to visibility of Http3SettingType

There used to be a build issue because of the visibility, but not anymore, so we can revert the change.

PatrikTrefil committed:
order Http3SettingType enum values in source code

PatrikTrefil committed:
remove api exclusion list

Not necessary anymore, because the API for which it was used was removed.

PatrikTrefil committed:
better method naming in MsQuicWebTransportSession

PatrikTrefil committed:
define values of enum options in ref WebTransport

PatrikTrefil committed:
improve WebTransportSessionCloseTests with assertions for application error codes on quic streams

PatrikTrefil committed:
refactor WebTransport tests code for writing capsules

PatrikTrefil committed:
refactors of capsule protocol in WebTransport tests

PatrikTrefil committed:
simplify exception throwing in WebTransport

CarnaViire commented on src/libraries/System.Net.Http/src/System/Net/WebTransport/MsQuicWebTransportSession.cs:
```diff
@@ -0,0 +1,897 @@
+    {
+        get
+        {
+            lock (SyncLock) { return field; }
```
If checked after state is closed, these should be already set, so no need for a lock

CarnaViire commented on src/libraries/System.Net.Http/src/System/Net/WebTransport/CapsuleProtocol/Capsules.cs:
```diff
@@ -0,0 +1,343 @@
+    }
+
+    /// <exception cref="CapsuleProtocolException">When the received length does not match the payload length</exception>
+    public static MaxBidirectionalStreamsCapsule Deserialize(ReadOnlyMemory<byte> buffer)
```
-> ReadOnlySpan? (nothing async)

CarnaViire commented on src/libraries/System.Net.Http/src/System/Net/WebTransport/MsQuicWebTransportSession.cs:
```diff
@@ -0,0 +1,897 @@
+        internal set
+        {
+            // HACK: make the value <= int.MaxValue so that we can use SemaphoreSlim
+            int newValue = value > int.MaxValue ? int.MaxValue : (int)value;
```
MsQuic limits the max number of stream to 65k so we can just throw/abort if it's bigger

CarnaViire commented on src/libraries/System.Net.Http/src/System/Net/WebTransport/MsQuicWebTransportExtendedConnectManager.cs:
```diff
@@ -0,0 +1,365 @@
+            {
+                if (_validationException != null)
+                {
+                    throw _validationException;
```
TODO: @CarnaViire look for ExceptionDispatchInfo

CarnaViire replied:
This can be an inner exception probably, so it won't have to be modified

PatrikTrefil replied:
@CarnaViire can you elaborate please? What do you mean this can be an inner exception?

PatrikTrefil committed:
prefer span over readonlymemory

PatrikTrefil committed:
simplify code

PatrikTrefil committed:
assert parameter name in argument exception

PatrikTrefil committed:
remove redundant null checks

This is already checked by compiler. Projects that have the compiler check turned off are not supported.

PatrikTrefil committed:
fix race condition in WT session close test

PatrikTrefil committed:
limit the WT session configuration to msquic limits

PatrikTrefil committed:
cleanup usings

PatrikTrefil committed:
fix race condition in WebTransport test

PatrikTrefil committed:
make the error message more generic

PatrikTrefil committed:
fix WebTransport session termination

PatrikTrefil committed:
properly handle bad headers in http3connection and webtransport

PatrikTrefil committed:
documentation on exception in WebTransportStream

CarnaViire commented on src/libraries/System.Net.WebTransport/src/System/Net/WebTransport/WebTransportStream.cs:
```diff
@@ -0,0 +1,156 @@
+    {
+        if (NetEventSource.Log.IsEnabled()) NetEventSource.Trace(this);
+
+        await DisposeAsyncCore().ConfigureAwait(false);
```
Dispose and DisposeAsync should do the same thing as much as possible (here DisposeAsyncCore is something that is only called from DisposeAsync)

PatrikTrefil replied:
The reason for that is that the derived class actually does something different for each. The derived class calls the `Dispose` method on the resources it holds in its own override of the `Dispose(bool)` method. The derived class calls the `DisposeAsync` method on the resources it holds in its own override of the `DisposeAsyncCore()` method. You can see these calls in `MsQuicWebTransportStream` class. Why it might seem confusing is that the `Stream` base class doesn't use the same pattern. My goal was to implement the recommendation from [this article](https://learn.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-disposeasync#implement-both-dispose-and-async-dispose-patterns). Does that make sense or you still see something wrong with the implementation?

CarnaViire commented on src/libraries/System.Net.WebTransport/src/System/Net/WebTransport/WebTransportStream.cs:
```diff
@@ -0,0 +1,156 @@
+
+        await DisposeAsyncCore().ConfigureAwait(false);
+
+        Dispose(false);
```
What is the reason for calling it this way?

PatrikTrefil replied:
See comment above for explanation.

CarnaViire commented on src/libraries/System.Net.Http/src/System/Net/WebTransport/MsQuicWebTransportStream.cs:
```diff
@@ -0,0 +1,593 @@
+    /// <inheritdoc/>
+    public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken = default)
+    {
+        CancellationTokenRegistration? ctr = RegisterCancellationCallback(QuicAbortDirection.Read, cancellationToken);
```
given that you're awaiting the ReadAsync result, I think this could be `async using CancellationTokenRegistration ctr` (I also think it should be possible to use a default value for "no registration" to avoid HasValue check?)

CarnaViire replied:
for the second point, e.g. https://github.com/dotnet/runtime/blob/bf17d45e7223638964621d28f62b4eb995286560/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs#L1270-L1273

PatrikTrefil replied:
good idea, thanks

CarnaViire commented on src/libraries/System.Net.WebTransport/src/System/Net/WebTransport/MsQuicWebTransportStream.cs:
```diff
@@ -0,0 +1,593 @@
+        }
+    }
+
+#pragma warning disable CA2016 // Forward the 'CancellationToken' parameter to methods
```
I'd add a short comment explaining the problem, at least linking the github issue

PatrikTrefil replied:
I have added a comment. Not sure which github issue you mean though.

CarnaViire replied:
https://github.com/dotnet/runtime/issues/121910

CarnaViire commented on src/libraries/System.Net.Http/src/System/Net/WebTransport/MsQuicWebTransportStream.cs:
```diff
@@ -0,0 +1,593 @@
+        }
+        catch (QuicException ex)
+        {
+            throw ExceptionHandler(ex);
```
A method's name should be a verb, otherwise it reads like a class name/constructor
So smth like `WrapException` or `GetException`
I'd avoid saying "handle exception" because "handle" is when the exception is not thrown any further

e.g.:
https://github.com/dotnet/runtime/blob/bf17d45e7223638964621d28f62b4eb995286560/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L1600

```
if (_abortException is not null)
{
    throw GetRequestAbortedException(_abortException);
}
```

or it could also incorporate throw inside like in https://github.com/dotnet/runtime/blob/bf17d45e7223638964621d28f62b4eb995286560/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs#L1343

PatrikTrefil replied:
I have changed the name, but I can't incorporate throw inside, because the compiler complains about it in some methods.

```cs
int F() {
    try { return CalculateF(); } catch (Exception e) { ThrowWrappedEx(e); }
}
```

Even if you make `ThrowWrappedEx` always throw, the complier complains that some codepath in `F` do not return an int value. I also tried marking the method with `DoesNotReturnAttribute`, but that does not fix it.

CarnaViire commented on src/libraries/System.Net.WebTransport/src/System/Net/WebTransport/MsQuicWebTransportSession.cs:
```diff
@@ -0,0 +1,904 @@
+
+    private readonly ReadOnlyMemory<byte> _idEncodedAsVariableLengthInteger;
+    private readonly QuicStream _connectStream;
+    private readonly IMsQuicWebTransportSessionConnectionManager _connectionManager;
```
Why is there an additional interface? we're already in the concrete (MsQuic) implementation `MsQuicWebTransportSession`, so this could be just the concrete class `MsQuicWebTransportSessionConnectionManager`?

PatrikTrefil replied:
The `MsQuicWebTransportExtendedConnectManager` class is used by `Http3Connection` and `MsQuicWebTransportSession`. The implementation is done in a way which makes the methods inteded to be called by the `MsQuicWebTransportSession` only visible in `MsQuicWebTransportSession` and the methods intended to be called by the `Http3Connection` only visible in `Http3Connection`. I thought this was a neat separation of the two usecases of the clases.

CarnaViire commented on src/libraries/System.Net.Http/src/System/Net/WebTransport/MsQuicWebTransportSession.cs:
```diff
@@ -0,0 +1,904 @@
+    private Channel<ChannelItem>? _pendingUnidirectionalStreams;
+    private Channel<ChannelItem>? _pendingBidirectionalStreams;
+
+    private List<MsQuicWebTransportStream>? _openStreams = [];
```
just curious why is it nullable if it's also initialized?

PatrikTrefil replied:
It's not nullable anymore, it was not necessary.

CarnaViire commented on src/libraries/System.Net.WebTransport/src/System/Net/WebTransport/MsQuicWebTransportExtendedConnectManager.cs:
```diff
@@ -0,0 +1,376 @@
+    {
+        if (NetEventSource.Log.IsEnabled()) NetEventSource.Trace(this, $"Rejecting stream {channelItem.QuicStream}.");
+
+        QuicStream stream = channelItem.QuicStream;
+        stream.Abort(QuicAbortDirection.Both, (long)Http3ErrorCode.WebTransportBufferedStreamRejected);
+        stream.Dispose();
+        channelItem.ArrayBuffer.Dispose();
```
this looks like smth that should have been in the ChannelItem's dispose?

PatrikTrefil replied:
You could argue that it should be in its dispose method, but the happy path is that `ChannelItem` holds the array buffer and quic stream while its waiting in the channel and those are then used to create a `WebTransportSession` instance. In this path the ChannelItem is not responsible for the disposal of these elements, so I don't know. To me it's not super clear and also this code is not anywhere else. But, if you are convinced that it's a better practice to put this code in the Dispose method let me know and I will update it.

PatrikTrefil committed:
fix typo in name of test

PatrikTrefil committed:
fix handling of server-initiated streams in Http3Connection

PatrikTrefil committed:
fix visibility of test method

accidentally a test method was private

PatrikTrefil committed:
handle server closing server-initiated stream very early

PatrikTrefil committed:
fix handling of unknown server-initiated streams in Http3Connection

PatrikTrefil committed:
hack for missing RESET_STREAM_AT support in H3 connection

PatrikTrefil committed:
improve exception handling

provide more details about the problem that caused the exception

PatrikTrefil committed:
improve documentation about the behavior of WT session termination

it's a note about not implementing an optional mechanism described in the RFC draft https://datatracker.ietf.org/doc/html/draft-ietf-webtrans-http3-12#section-6-10

PatrikTrefil committed:
improve test and add note about it being flaky

PatrikTrefil committed:
simplify test OpeningStreamBeforeSessionExistsWorks

PatrikTrefil committed:
add more WT session establishment tests

PatrikTrefil committed:
improve code readability

PatrikTrefil committed:
add missing dispose calls in case of exceptions in MsQuicWebTransportExtendedConnectManager.Process.ReceivedStreamAsync

PatrikTrefil committed:
improve documentation

PatrikTrefil committed:
fix double return of array to array pool in ConcatenatedStream

PatrikTrefil committed:
fix ObjectDisposedException in MsQuicWebTransportStream

one should never call QuicStream.Dispose and QuicStream.Abort in parallel

PatrikTrefil committed:
use ArrayBuffer in CapsuleConsumer and fix double return of array to array pool

PatrikTrefil committed:
cleanup usings

PatrikTrefil committed:
fix race condition in msquic webtransport session

calling abort during dispose on QuicStream was causing issues

PatrikTrefil committed:
add tracing for exception in MsQuicWebTransportExtendedConnectManager

PatrikTrefil committed:
remove redundant lock as optimization

PatrikTrefil committed:
fix value of property AllowDuplex in Http3ExtendedConnectContent

we are not actually doing duplex communication

PatrikTrefil committed:
fix race condition in test

PatrikTrefil committed:
improve code quality and add comments in MsQuicWebTransportStream

PatrikTrefil committed:
improve naming in MsQuicWebTransportStream

PatrikTrefil committed:
add missing object disposed checks to MsQuicWebTransportStream

PatrikTrefil committed:
simplify handling of open streams in MsQuicWebTransportSession

PatrikTrefil committed:
increase timeout of http requests in WT tests

was failing because of default timeout 3 s

PatrikTrefil committed:
add missing throw

PatrikTrefil committed:
fix race condition in test

the short timeout was unintentionally affecting the setup of the background session, which was timing out on slow computers

PatrikTrefil committed:
fix handling of exceptions in ClientWebTransportSession

PatrikTrefil committed:
fix typos

PatrikTrefil committed:
set isDisposed to true in DisposeAsyncCore

PatrikTrefil committed:
add comment explaining synchro

PatrikTrefil committed:
fix and improve docs

PatrikTrefil committed:
fix typos

PatrikTrefil committed:
fix not returning array to array pool in case of error

PatrikTrefil committed:
fix formatting

PatrikTrefil committed:
match api in ref to the actual api

PatrikTrefil committed:
align ref with actual implementation

PatrikTrefil committed:
use properties rather than fields

PatrikTrefil committed:
add missing dispose in case of error

PatrikTrefil committed:
add missing exception messages

PatrikTrefil committed:
fix method call

PatrikTrefil committed:
fix lazy init in DrainSessionCapsule

PatrikTrefil committed:
fix constant in CloseSessionCapsule

the value was in bits, but was supposed be in bytes

PatrikTrefil committed:
fix exception handling in ClientWebTransportSession

PatrikTrefil committed:
split webtransport into separate project

PatrikTrefil committed:
split unit and functional tests into separate projects

PatrikTrefil committed:
clean up API, fix RFC compliance by System.Net.Http and unify neteventsource in WebTransport

PatrikTrefil committed:
fix method call to match changes in last commit

PatrikTrefil committed:
add missing docstrings

PatrikTrefil committed:
add tests for h/3 extended connect

PatrikTrefil committed:
fix fire and forget in h/3 connection go away processing

PatrikTrefil committed:
add remarks to method for better documentation

PatrikTrefil committed:
fix disposal of resources in Http3Connection and test

PatrikTrefil committed:
fix version negotiation in ClientWebTransportSession

if the user requested H/2 and RequestVersionHigher, the handling was incorrect

PatrikTrefil committed:
fix Http3Connection.OpenOutboundStreamAsync when _pool.Settings.EnableMultipleHttp3Connections == true

PatrikTrefil committed:
fix tombstoning of buffered streams

PatrikTrefil committed:
remove already fixed todo

PatrikTrefil committed:
fix name of field

PatrikTrefil committed:
fix stream errors codes during connection abort

PatrikTrefil committed:
fix OpenOutboundStreamAsyncCore implementation in MsQuicWebTransportSession

PatrikTrefil committed:
rework stream clean up

PatrikTrefil committed:
remove unnecesssary usings

PatrikTrefil committed:
simplify name

PatrikTrefil committed:
fix and simplify exception handling

PatrikTrefil committed:
complete and close pending streams in background

PatrikTrefil committed:
simplify accessibility

PatrikTrefil committed:
fix stream limit accounting in Http3Connection

works, but suboptimal, because all waiters are woken up when a stream becomes available

PatrikTrefil committed:
fix request hanging until timeout when server sends no settings

PatrikTrefil committed:
improve code readability

PatrikTrefil committed:
fix handling of GOAWAY in WebTransport

PatrikTrefil committed:
fix build api compat issues

PatrikTrefil committed:
add unit tests to WebTransport solution

PatrikTrefil committed:
add WebTransport networking events to TestEventListener.NetworkingEvents

PatrikTrefil committed:
add missing early return in ValidateAndProcessServerSettings

PatrikTrefil committed:
add fallback to CapsuleSenderExceptionHandler

PatrikTrefil committed:
fix and improve exception handling in MsQuicWebTransportSession

PatrikTrefil committed:
simplify code

PatrikTrefil committed:
fix MsQuicWebTransportStream.Dispose

PatrikTrefil committed:
fix missing cancellation token pass

PatrikTrefil committed:
fix formatting

PatrikTrefil committed:
fix link to rfc

PatrikTrefil committed:
fix max data sent limit handling

PatrikTrefil committed:
fix confusing comments

PatrikTrefil committed:
fix comment

PatrikTrefil committed:
fix typos

PatrikTrefil committed:
add null checks as recommended in the guidelines

PatrikTrefil committed:
fix documentation in WebTransportSession

PatrikTrefil committed:
fix documentation

PatrikTrefil committed:
fix MsQuicWebTransportSession cleanup

we were not disposing the clean up semaphore

PatrikTrefil committed:
remove todo out of scope

performance testing is out of scope

PatrikTrefil committed:
remove already fixed todo

PatrikTrefil committed:
remove todo - won't do

PatrikTrefil committed:
add link to issue to comment

PatrikTrefil committed:
add missing docstrings to public api members

PatrikTrefil committed:
remove unused property

leftover from previous approach

PatrikTrefil committed:
fix malformed xml in docstring

PatrikTrefil committed:
add missing param tags to docstring

PatrikTrefil committed:
fix typos

PatrikTrefil committed:
fix docstring

PatrikTrefil committed:
fix formatting

PatrikTrefil committed:
add tracing to sending data in wt

PatrikTrefil committed:
fix tracing in capsule protocol

PatrikTrefil committed:
fix trace level in webtransport

PatrikTrefil committed:
fix zero byte reads

PatrikTrefil committed:
split session manager and extended connect manager
